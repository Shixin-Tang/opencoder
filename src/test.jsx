import React, { useState, useEffect } from "react"
import { render, Box, Text } from "ink2"
import Static from "./components/static.js"
import { render as renderInk } from "../test/testutil"

const Example = () => {
  const [tests, setTests] = useState([])

  useEffect(() => {
    let completedTests = 0
    let timer

    const run = () => {
      // Fake 10 completed tests
      if (completedTests++ < 10) {
        setTests((previousTests) => [
          ...previousTests,
          {
            id: previousTests.length,
            title: `Test #${previousTests.length + 1}`,
          },
        ])

console.log(renderInk(<Text color="green">✔Hello</Text>).lastFrame())

        timer = setTimeout(run, 100)
      }
    }

    run()

    return () => {
      clearTimeout(timer)
    }
  }, [])

  return (
    <Box flexDirection="column" gap={1} position="relative">
      {/* This part will be rendered once to the terminal */}
      <Static items={tests}>
        {(test) => (
          <Box key={test.id}>
            <Text color="green">✔ {test.title}</Text>
          </Box>
        )}
      </Static>

      {/* This part keeps updating as state changes */}
      <Box marginTop={1}>
        <Text dimColor>Completed tests: {tests.length}</Text>
      </Box>
    </Box>
  )
}

render(<Example />)
