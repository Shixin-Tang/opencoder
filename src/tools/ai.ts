import type { ReactNode } from "react"
import { type JSONValue, type Tool, type ToolInvocation } from "ai"

// TODO: annotate message with tool output params, eg file patch, useful for persistent chat messages.

type Streamable = ReactNode | Promise<ReactNode> | JSONValue

export type Renderer<T extends Array<any>> = (...args: T) => AsyncGenerator<Streamable, void, void>

export type CoderTool = Omit<Tool, "execute"> & {
  generate?: Renderer<any>
  renderTitle?: (part: ToolInvocation) => ReactNode
  render?: (part: ToolInvocation) => ReactNode
  execute?: Tool["execute"]
}

export function defineTool(params: CoderTool): CoderTool {
  return params
}
