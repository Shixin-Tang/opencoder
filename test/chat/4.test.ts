import { type LanguageModelV1StreamPart } from "ai"
import { convertArrayToReadableStream, MockLanguageModelV1 } from "ai/test"

import { waitNextRender } from "../utils/render.js"
import { setupTestEnvironment } from "./util.js"

test("handling model stream error during response generation", async () => {
  // 1. Setup model to stream partial response then error
  const errorStreamChunks: LanguageModelV1StreamPart[] = [
    { type: "text-delta", textDelta: "This is the first part. " },
    { type: "text-delta", textDelta: "Then things go wrong... " },
    // Simulate an error from the stream
    { type: "error", error: "Something went wrong during generation!" },
    // Note: A 'finish' chunk might or might not arrive after an error,
    // depending on the provider. We simulate without one here.
  ]

  const mockModel = new MockLanguageModelV1({
    doStream: async (options) => ({
      stream: convertArrayToReadableStream(errorStreamChunks),
      rawCall: { rawPrompt: JSON.stringify(options.prompt), rawSettings: {} },
    }),
  })

  const { instance, stdin, stdout, fiber } = await setupTestEnvironment({ model: mockModel })
  assert(stdin)

  expect(stdout.get()).toMatchSnapshot("stream error - initial")

  // 2. Simulate user input
  stdin.emit("input", "Trigger the error stream")
  await waitNextRender()

  // Restore waitFor to allow time for error processing and input re-enabling
  await vi.waitFor(
    () => {
      const output = stdout.get()
      // Verify input prompt is showing again
      expect(output).toContain("> ")
    },
    { timeout: 2000 },
  )

  // 4. Final state snapshot
  expect(stdout.get()).toMatchSnapshot("stream error - final state")

  // 5. Ensure app remains interactive
  stdin.emit("input", " can I type again?")
  await waitNextRender()
  expect(stdout.get()).toContain("can I type again?")

  instance.unmount()
})
