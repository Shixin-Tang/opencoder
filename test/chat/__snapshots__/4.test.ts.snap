// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`handling model stream error during response generation > stream error - final state 1`] = `
"
╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│ > Trigger the error stream                                                                       │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
  / for commands                                                     alt + ⏎ for newline ·0 tokens"
`;

exports[`handling model stream error during response generation > stream error - initial 1`] = `
"
╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│ >                                                                                                │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
  / for commands                                                     alt + ⏎ for newline ·0 tokens"
`;
